### PR TITLE
Structure is the wrong size(0x2d5) issue with tpm2_nvdefine without specifying the size

### PR DESCRIPTION
**Pull request by Ivanhu5866**
Please refer to
https://github.com/tpm2-software/tpm2-tools/issues/2171

### CODE REVIEW

File: tools/tpm2_nvdefine.c [hunk 1]
```diff
@@ -201,27 +200,50 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
...
+
+    TPM2_NV_INDEX max = 0;
+    UINT16 nv_max_sz = 0;
+
+    tool_rc tmp_rc = get_max_nv_index(ectx, &nv_max_sz);
+    if (tmp_rc != tool_rc_success) {
+        return tmp_rc;
+    }
+
+    max = TPM2_HR_NV_INDEX | nv_max_sz;
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:239]: bitwise or, why? Why are you adding the handle mask into the max size of an NV index? max will always be set and code on 240 will be dead. The original code checked that nv_max_size wasn't 0 before adding in this part of the handle, but thats for handles, not for size.

File: tools/tpm2_nvdefine.c [hunk 2]
```diff
@@ -183,11 +183,10 @@ static void handle_default_attributes(void) {
     }
 }
 
-static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chosen) {
+static tool_rc get_max_nv_index(ESYS_CONTEXT *ectx, UINT16 *size) {
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:186]: get_max_nv_index_size

File: tools/tpm2_nvdefine.c [hunk 3]
```diff
@@ -201,27 +200,50 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
     UINT32 count = capabilities->data.tpmProperties.count;
 
     if (!count) {
-        LOG_ERR("Could not get maximum NV index, try specifying an NV index");
+        LOG_ERR("Could not get maximum NV index");
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:203]: LOG_ERR("Could not get maximum NV index size");

File: tools/tpm2_nvdefine.c [hunk 4]
```diff
@@ -271,12 +294,17 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
...
-static tool_rc validate_size(void) {
+static tool_rc validate_size(ESYS_CONTEXT *ectx) {
+
+    tool_rc rc = tool_rc_general_error;
 
     switch ((ctx.nv_attribute & TPMA_NV_TPM2_NT_MASK) >> TPMA_NV_TPM2_NT_SHIFT) {
         case TPM2_NT_ORDINARY:
             if (!ctx.size_set) {
-                ctx.size = TPM2_MAX_NV_BUFFER_SIZE;
+		if ((rc = get_max_nv_index(ectx, &ctx.size)) != tool_rc_success) {
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:304]: we don't hid functions call sin conditionals unless it makes the code simpler, this doesn't, so:
```C
rc = get_max_nv_index(ectx, &ctx.size);
if (rc != tool_rc_success) {
...
}
```

File: tools/tpm2_nvdefine.c [hunk 5]
```diff
@@ -271,12 +294,17 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
...
+			ctx.size = TPM2_MAX_NV_BUFFER_SIZE;
+			return rc;
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:306]: So fail validate_size but set the ctx.size to max? I don't even think that will work as callers of validate_size will cause the tool to exit. I think you want to return tool_rc_success

File: tools/tpm2_nvdefine.c [hunk 6]
```diff
@@ -183,11 +183,10 @@ static void handle_default_attributes(void) {
     }
 }
 
-static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chosen) {
+static tool_rc get_max_nv_index_size(ESYS_CONTEXT *ectx, UINT16 *size) {
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:186]: make this void, the return code isn't really used in any meaningful and consistent way. In one path, its only use is to set the default max size, in the other path it causes the tool to exit. Just make this on error set the *size to max via the define and print a warning.

File: tools/tpm2_nvdefine.c [hunk 7]
```diff
@@ -201,27 +200,47 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
...
+
+    tool_rc rc = tool_rc_general_error;
+
+    TPM2_NV_INDEX max = 0;
+    UINT16 nv_max_sz = 0;
+
+    tool_rc tmp_rc = get_max_nv_index_size(ectx, &nv_max_sz);
+    if (tmp_rc != tool_rc_success) {
+        return tmp_rc;
+    }
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:238]: You shouldn't need to modify handle_no_index_specified(). It's just looking up the first free NV index, it has nothing to do with size.

File: tools/tpm2_nvdefine.c [hunk 8]
```diff
@@ -201,27 +200,47 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
...
+
+    max = TPM2_HR_NV_INDEX + nv_max_sz;
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:240]: This makes no sense to me? Why would you take the maximum size of an nv index in bytes, turn it into an NV index, assign it to max, and then never use max again?

File: tools/tpm2_nvdefine.c [hunk 9]
```diff
@@ -201,27 +200,47 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
...
+    rc = tool_rc_success;
+
+out:
     free(capabilities);
-    capabilities = NULL;
+
+    return rc;
+}
+
+static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chosen) {
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:228]: leave this function alone.

File: tools/tpm2_nvdefine.c [hunk 10]
```diff
@@ -271,12 +291,17 @@ static tool_rc handle_no_index_specified(ESYS_CONTEXT *ectx, TPM2_NV_INDEX *chos
...
-static tool_rc validate_size(void) {
+static tool_rc validate_size(ESYS_CONTEXT *ectx) {
+
+    tool_rc rc = tool_rc_general_error;
 
     switch ((ctx.nv_attribute & TPMA_NV_TPM2_NT_MASK) >> TPMA_NV_TPM2_NT_SHIFT) {
         case TPM2_NT_ORDINARY:
             if (!ctx.size_set) {
-                ctx.size = TPM2_MAX_NV_BUFFER_SIZE;
+		rc = get_max_nv_index_size(ectx, &ctx.size);
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:301]: just make this void and set on failure to TPM2_MAX_NV_BUFFER_SIZE and print a warning on this case.

File: tools/tpm2_nvdefine.c (hunk 6 above)
Comment (williamcroberts) [reply tools/tpm2_nvdefine.c:186]: In fact the error case where it causes the tool to exit shouldn't even exist. Their should be no changes to handle_no_index_specified AFAICT.

Comment (Ivanhu5866) [reply tools/tpm2_nvdefine.c:186]: Ok...will make get_max_nv_index_size simpler without modifying handle_no_index_specified.

File: tools/tpm2_nvdefine.c (hunk 7 above)
Comment (Ivanhu5866) [reply tools/tpm2_nvdefine.c:238]: Ok, will skip this patch.

File: tools/tpm2_nvdefine.c (hunk 8 above)
Comment (Ivanhu5866) [reply tools/tpm2_nvdefine.c:240]: I think max will be used for the range to fine the first nv index, maybe I am wrong, but it doesn't matter, this patch will be skipped.

File: tools/tpm2_nvdefine.c (hunk 10 above)
Comment (Ivanhu5866) [reply tools/tpm2_nvdefine.c:301]: OK

File: tools/tpm2_nvdefine.c (hunk 9 above)
Comment (Ivanhu5866) [reply tools/tpm2_nvdefine.c:228]: Ok

**Comment by idesai** [comment]
> Please refer to
> #2171

Fixes #2171

File: tools/tpm2_nvdefine.c [hunk 11]
```diff
@@ -183,6 +183,48 @@ static void handle_default_attributes(void) {
...
+	*size = TPM2_MAX_NV_BUFFER_SIZE;
+        LOG_ERR("Could not get fixed TPM properties");
+        return;
+    }
+
+    TPMS_TAGGED_PROPERTY *properties = capabilities->data.tpmProperties.tpmProperty;
+    UINT32 count = capabilities->data.tpmProperties.count;
+
+    if (!count) {
+	*size = TPM2_MAX_NV_BUFFER_SIZE;
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:204]: indent error, tab. We use 4 spaces.

File: tools/tpm2_nvdefine.c [hunk 12]
```diff
@@ -183,6 +183,48 @@ static void handle_default_attributes(void) {
...
+    UINT32 i;
+    for (i=0; i < count; i++) {
+        if (properties[i].property == TPM2_PT_NV_INDEX_MAX) {
+		*size = properties[i].value;
+		break;
+        }
+    }
+
+    if (*size == 0) {
+	*size = TPM2_MAX_NV_BUFFER_SIZE;
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:218]: indent error, tab, use four spaces

File: tools/tpm2_nvdefine.c [hunk 13]
```diff
@@ -183,6 +183,48 @@ static void handle_default_attributes(void) {
...
+static void get_max_nv_index_size(ESYS_CONTEXT *ectx, UINT16 *size) {
+
+    *size = 0;
+
+    /* get the max NV index for the TPM */
+    TPMS_CAPABILITY_DATA *capabilities = NULL;
+    tool_rc tmp_rc = tpm2_getcap(ectx, TPM2_CAP_TPM_PROPERTIES, TPM2_PT_FIXED,
+            TPM2_MAX_TPM_PROPERTIES, NULL, &capabilities);
+    if (tmp_rc != tool_rc_success) {
+	*size = TPM2_MAX_NV_BUFFER_SIZE;
```
Comment (williamcroberts) [inline tools/tpm2_nvdefine.c:195]: Here to, spaces instead of tab

**Pull request merged**